tools: Make clang-format-diff.py use versioned files, not local files

aghiti opened the pull request:
clang-format-diff.py uses the 'diff' passed in stdin to know which files
to pass to clang-format. clang-format then uses the *local* versions of those
files to check any formatting issues: then if formatting is correct locally
but not on HEAD (which is currently being pushed), the script does not
complain and then we can push unformatted source files.

Fix that by checkouting those files into /tmp directory and make clang-format
work on the temp files.

aghiti committed:
tools: Make clang-format-diff.py use versioned files, not local files

clang-format-diff.py uses the 'diff' passed in stdin to know which files
to pass to clang-format. clang-format then uses the *local* versions of those
files to check any formatting issues: then if formatting is correct locally
but not on HEAD (which is currently being pushed), the script does not
complain and then we can push unformatted source files.

Fix that by checkouting those files into /tmp directory and make clang-format
work on the temp files.

aghiti commented:
The PR is in clang to have a diff, this new version will end up in upmem_utilities.

vpalatin commented on tools/clang-format/clang-format-diff.py:
```diff
@@ -65,11 +66,24 @@ def main():
+                         stderr=None,
+                         stdin=subprocess.PIPE,
+                         universal_newlines=True)
+      stdout, stderr = p.communicate()
```
this is the complicated version of subprocess.check_output() ?

vpalatin commented on tools/clang-format/clang-format-diff.py:
```diff
@@ -114,15 +132,17 @@ def main():
     if not args.i:
-      with open(filename) as f:
-        code = f.readlines()
+      tmpfile.seek(0)
```
if you no longer open the file/keep the common path, then you have already recorded the content, you can pass it directly rather than reading from the disk

aghiti replied:
clang-format (called at line 125) needs the file name anyway. I can also pass the content in the tuple.

vpalatin commented on tools/clang-format/clang-format-diff.py:
```diff
@@ -114,15 +132,17 @@ def main():
       formatted_code = StringIO(stdout).readlines()
       diff = difflib.unified_diff(code, formatted_code,
-                                  filename, filename,
+                                  filename, tmpfile.name,
```
I'm surprised that you are still using 'filename' here, isn't the (potentially modified) local copy not what you want ?

aghiti replied:
Those two fields are only here to print the filenames, the diff being made from the `code` and `formatted_code` buffers. I left one `filename` so that the user knows which file it is
while still knowing that it is not the local file that is compared:

```
--- api/src/api/dpu.c	(before formatting)
+++ /tmp/tmpLPiTgF	(after formatting)
```

rather than having:

```
--- /tmp/tmpLPiTgF	(before formatting)
+++ /tmp/tmpLPiTgF	(after formatting)
```

vpalatin replied:
then you want both side to be filename ?

aghiti committed:
- add revision argument
- add extension file
- add .clang-format checkout
- pass file content directly
- use check_output when possible

rjodinupmem commented on tools/clang-format/clang-format-diff.py:
```diff
@@ -65,11 +66,24 @@ def main():
+      # Do not use the local file but the version of the file that
+      # is committed in HEAD.
       filename = match.group(2)
+      command = ("git show HEAD:" + filename).split(" ")
```
``HEAD`` feels weird here. What happened if your are not pushing the branch actually checkout ? Maybe we would like to have a optional argument to clang-format-diff set to ``HEAD`` by default.

aghiti replied:
Agreed, I had not thought of this case. I'll add the argument.

aghiti commented:
> To respect the convention on LLVMs repositories, please change your commit message so that it is prefix by `dpu: generic:`.
> This will ease the process of moving to release 9 soon
> Thanks

As said in previous comment, I did not want to push the new version of this script into clang. Unless you want it here

ghost commented:
> > To respect the convention on LLVMs repositories, please change your commit message so that it is prefix by `dpu: generic:`.
> > This will ease the process of moving to release 9 soon
> > Thanks
> 
> As said in previous comment, I did not want to push the new version of this script into clang. Unless you want it here

Sounds good to me

vpalatin commented on tools/clang-format/clang-format-diff.py:
```diff
@@ -103,13 +107,14 @@ def main():
-  for filename_dir, lines in lines_by_file.items():
-    filename = filename_dir[0]
-    tmpfile = filename_dir[1]
+  for file_tuple, lines in lines_by_file.items():
```
style note: you can write
```for (filename, tmpfile, code), lines in lines_by_file.items():```

aghiti replied:
Ok thx.

aghiti replied:
Regarding "then you want both side to be filename ?": I don't understand

vpalatin replied:
> Regarding "then you want both side to be filename ?": I don't understand

Your answer is in a strange place/thread.
If the 2 fields in the difflib.unified_diff() call are only used to *display* the filenames I don't really want any '/tmp/badToto.tmp' , both of them should be the real filename ?

aghiti replied:
Ok.

aghiti committed:
- both filename

aghiti commented:
Merged in upmem_utilities.
